// common/responseTool.js
var responseTool = {
    tool: require('./tool'),
    jsdom: require('../public_nodejs/jsdom').jsdom,
    fs: require('fs'),
    path: require('path'),
    /*返回服务端异常*/
    sendServerException: function (res, mess) {
        res.send(500, mess || '服务端错误请查看后台日志');
    },
    /*返回拒绝访问*/
    sendDecline: function (res, mess) {
        res.send(405, mess || '服务端拒绝访问');
    },
    /*返回数据更新、上传、下载成功*/
    sendSuccess: function (res, param) {
        param = param || {};
        res.set('Content-Type', 'text/json;charset=utf-8');
        res.send(param);
    },
    /*重定向*/
    sendRedirect: function (res, url) {
        res.send(302, url);
    },
    /*登录请求的响应
     */
    loginResponse: function (req, res, result) {
        var err = result.err;
        if (err) {
            console.error('登录err：' + (err.stack || JSON.stringify(err)));
            return this.renderLogin(res, err, result.loginType);
        }
        this.tool.storeUserInfo(req, res, result.puser);
        this.redirectMain(res);
    },
    /*渲染主页面*/
    renderMain: function (res, userInfo, ticket) {
        var _this = this;
        var mainPath = './views/';
        var mainHtml = 'layout.html';
        var mainHtmlPath = mainPath + mainHtml;

        var newMainHtmlName = '_layout.html';
        var newMainHtmlPath = mainPath + newMainHtmlName;
        //var newMainHtmlName = Math.random() + '.html';
        //var newMainHtmlPath = _this.path.join(uploadPath, newMainHtmlName);

        _this.fs.readFile(mainHtmlPath, 'utf8', function (readErr, readResult) {
            if (readErr) {
                console.error('读取layout err：' + readErr);
                return _this.sendDecline(res);
            }
            readResult = '<script type="text/javascript">var ' + pconst.pticket + '="' + ticket + '";</script>' + readResult;
            _this.fs.writeFile(newMainHtmlPath, readResult, 'utf8', function (writeErr, writeResult) {
                if (writeErr) {
                    console.error('写入layout err：' + writeErr);
                    return _this.sendDecline(res);
                }
                res.render(newMainHtmlName, { host: commonLibUrl, user: userInfo });
            });
        });
    },
    /*渲染登录页面*/
    renderLogin: function (res, err, loginType) {
        res.render('login', {
            host: commonLibUrl,
            err: err || '',
            loginType: parseInt(loginType) || 1
        });
    },
    /*跳转到首页*/
    redirectMain: function (res) {
        res.redirect('/');
    },
    /*无权限访问*/
    renderNoPower: function (res, isLogin) {
        if (isLogin === true) return this.renderLogin(res);
        this.sendDecline(res, '无权限访问');
    }
};

module.exports = responseTool;
